rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp

```diff
@@ -0,0 +1,220 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__LIFECYCLE_NODE_HPP_
+#define RCLCPP__LIFECYCLE_NODE_HPP_
+
+#include "rclcpp/node.hpp"
+#include "rclcpp_lifecycle/lifecycle_publisher.hpp"
+
+#include "rcl_lifecycle/lifecycle_state.h"
+#include "rcl_lifecycle/default_state_machine.h"
+#include "rcl_lifecycle/transition_map.h"
+
+namespace rclcpp
+{
+namespace node
+{
+
+namespace lifecycle_interface
+{
+
+/**
+ * @brief Interface class for a managed node.
+ * Pure virtual functions as defined in
+ * http://design.ros2.org/articles/node_lifecycle.html
+ */
+class NodeInterface
+{
+  virtual bool on_configure()   = 0;
+  virtual bool on_clean_up()    = 0;
+  virtual bool on_shutdown()    = 0;
+  virtual bool on_activate()    = 0;
+  virtual bool on_deactivate()  = 0;
+  virtual bool on_error()       = 0;
+};
+}  // namespace lifecycle_interface
+
+#include <rclcpp_lifecycle/visibility_control.h>
+
+/**
+ * @brief LifecycleNode as child class of rclcpp Node
+ * has lifecycle nodeinterface for configuring this node.
+ */
+class LifecycleNode : public rclcpp::node::Node, public lifecycle_interface::NodeInterface
```

Comment:
This is starting to show the problem of not having a `Node` interface. "Favor 'object composition' over 'class inheritance'" is often thrown around and abused, but C++ best practice generally says to not inherit implementations except for minor tweaks to behaviour. It may work as is for this case, but for a major entity like Node, having an abstract interface would be more flexible, and would make things more testable.